### ISSUE TITLE
Non-fatal errors responses for `presence.enter()` cause Publisher to throw exceptions

### ISSUE THREAD

**Issue by jaley** [ably/ably-asset-tracking-android#907]
The publisher treats non-fatal errors from `presence.enter()` (which can be retried) as fatal during calls to `track()`. This is reproduced in `faultBeforeAddingTrackable[EnterFailedWithNonfatalNack]` on #886. Initial analysis below:

## faultBeforeAddingTrackable[EnterFailedWithNonfatalNack]

Publisher `track()` call fails because the `presence.enter()` call fails when ably-java calls the
`onError` handler in response to the injected (non-fatal) `NACK`. This results in an exception and
the operation is aborted with a `ConnectionException` thrown to the caller. This is a non-fatal error,
which should result in transparent retries in the background, with the trackable in the Offline state
until it succeeds.

```
01-13 15:15:39.073 19291 20859 D PUBLISHING SDK IT: 19729:  13-01-23 15:15:39.072: [DefaultAbly] Failed to suspend enter presence for channel com.ably.tracking.common.DefaultAblySdkRealtime$Channel@23e3f6d.name com.ably.tracking.ConnectionException: injected by proxy
01-13 15:15:39.075 19291 20048 D PUBLISHING SDK IT: 19318:  13-01-23 15:15:39.074: [DefaultAbly] Failed to enter the presence of channel tracking:58383c6a-0535-48b0-a19f-097b81707b47 com.ably.tracking.ConnectionException: injected by proxy
01-13 15:15:39.076 19291 20048 D PUBLISHING SDK IT: 19318:  13-01-23 15:15:39.075: [DefaultAbly] Failed to connect for channel tracking:58383c6a-0535-48b0-a19f-097b81707b47 com.ably.tracking.ConnectionException: injected by proxy
01-13 15:15:39.079 19291 20048 D PUBLISHING SDK IT: 19318:  13-01-23 15:15:39.078: [DefaultAbly] Failed to connect for trackable 58383c6a-0535-48b0-a19f-097b81707b47 com.ably.tracking.ConnectionException: injected by proxy
01-13 15:15:39.082 19291 19398 D PUBLISHING SDK IT: 19040:  13-01-23 15:15:39.081: [DefaultPublisher] Publisher track operation failed com.ably.tracking.ConnectionException: injected by proxy
01-13 15:15:39.082 19291 20048 D PUBLISHING SDK IT: 19318:  Layer7Proxy: (raw) [ClientToServer] {"action":12,"channel":"tracking:58383c6a-0535-48b0-a19f-097b81707b47"}
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 19007:  attempt to add active Trackable while fault active - failed - com.ably.tracking.ConnectionException: injected by proxy
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: com.ably.tracking.ConnectionException: injected by proxy
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at com.ably.tracking.common.AblyHelpersKt.toTrackingException(AblyHelpers.kt:61)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at com.ably.tracking.common.DefaultAbly$enterPresenceSuspending$2$1.onError(Ably.kt:922)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at io.ably.lib.transport.ConnectionManager$PendingMessageQueue.nack(ConnectionManager.java:1701)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at io.ably.lib.transport.ConnectionManager.onNack(ConnectionManager.java:1273)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at io.ably.lib.transport.ConnectionManager.onMessage(ConnectionManager.java:1161)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at io.ably.lib.transport.WebSocketTransport$WsClient.onMessage(WebSocketTransport.java:183)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at org.java_websocket.client.WebSocketClient.onWebsocketMessage(WebSocketClient.java:641)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at org.java_websocket.drafts.Draft_6455.processFrameBinary(Draft_6455.java:959)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at org.java_websocket.drafts.Draft_6455.processFrame(Draft_6455.java:912)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at org.java_websocket.WebSocketImpl.decodeFrames(WebSocketImpl.java:401)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at org.java_websocket.WebSocketImpl.decode(WebSocketImpl.java:233)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at org.java_websocket.client.WebSocketClient.run(WebSocketClient.java:516)
01-13 15:15:39.083 19291 19319 D PUBLISHING SDK IT: 	at java.lang.Thread.run(Thread.java:1012)
```

**Comment by lawrence-forooghian** [comment]
@jaley does this not fall under the umbrella of #871? Specifically (emphasis mine):

> If any of the operations required to add and track failed for a non-retriable reason, then it's `Failed`. If all of the operations required are in one of the following states:
> 
> * still in progress; or
> * succeeded; or
> * *failed for a retriable reason*
> 
> *then it's `Offline`.*

I haven't looked through @davyskiba’s PR #883 — which claims to resolve #871 — in detail yet. But it seems that this test is now passing on that branch.

**Comment by paddybyers** [comment]
I think many of these problems are related and ultimately we might find that we can merge multiple issues as having the same underlying cause.

However, I do think there are multiple specific problems:

a) channel lifecycle events - attach/detach - are not handled correctly when they fail;
b) channel publish operations - publish of regular messages, plus presence enter/leave - are not handled correctly when they fail.

The confusion is coming, at least in part, from the fact that both a channel `attach()` and a presence `enter()` are involked in a call to `track()`.

So I think it would be good if we could characterise the specific underlying operations that are being attempted, how we expect them to behave, and how they're actually behaving. And them some of the issues we find will be associated with AAT, and some with ably-java.

So lets progress each of these issues to the point that we can characterise which operations are behaving incorrectly. If, once we know that, we find that the issues are the same, then we can close one as a duplicate.

If we can do it, it might be worth trying construct integration tests that isolate failures in the (a) and (b) operations independently. We can isolate (b) events, by injecting a fault after `track()` has succeeded, and seeing what happens to the publishes associated with location updates (ie nothing to do with presence).

We do have faults already I think that model (a) events failing.

**Comment by deanna-lad** [comment]
The link to the corresponding Jira issue is https://ably.atlassian.net/browse/SDK-3245

**Comment by lawrence-forooghian** [comment]
This one is failing because `DefaultAbly.presence.enter()` is throwing a `ConnectionException` in response to the received `ProtocolMessage(NACK)` causing the ably-java presence enter operation to fail. @davyskiba is going to add retry logic for this case to #883, so I’ll assign this one to him.

**Comment by davyskiba** [comment]
Resolved by #938
